ngcplogger.go

```diff
@@ -345,6 +362,97 @@ func (l *nGCPLogger) extractMsgFromPayload(m map[string]any) {
 	}
 }
 
+func (l *nGCPLogger) extractGcpFromPayload(m map[string]any, entry *logging.Entry) {
+
+	if l.extractGcp {
+		if val, exists := m["logging.googleapis.com/sourceLocation"]; exists {
+			v := val.(map[string]any)
+			entry.SourceLocation = &loggingpb.LogEntrySourceLocation{
+				File:     v["file"].(string),
+				Line:     int64(v["line"].(float64)),
+				Function: v["function"].(string),
+			}
+			delete(m, "logging.googleapis.com/sourceLocation")
+		}
+		if val, exists := m["logging.googleapis.com/trace"]; exists {
+			entry.Trace = val.(string)
+			delete(m, "logging.googleapis.com/trace")
+		}
+		if val, exists := m["logging.googleapis.com/spanId"]; exists {
+			entry.SpanID = val.(string)
+			delete(m, "logging.googleapis.com/spanId")
+		}
+		if val, exists := m["logging.googleapis.com/trace_sampled"]; exists {
+			entry.TraceSampled = val.(bool)
+			delete(m, "logging.googleapis.com/trace_sampled")
+		}
+		if val, exists := m["logging.googleapis.com/labels"]; exists {
+			v := val.(map[string]any)
+			for k, v := range v {
+				entry.Labels[k] = v.(string)
+			}
+			delete(m, "logging.googleapis.com/labels")
+		}
+	}
+}
+
+func (l *nGCPLogger) extractCaddyFromPayload(m map[string]any, entry *logging.Entry) {
+
+	if l.extractCaddy {
+		if val, exists := m["request"]; exists {
+			hr := logging.HTTPRequest{
+				Request: &http.Request{
+					Header: make(http.Header),
+				},
+			}
+			v := val.(map[string]any)
+			hr.Request.Method = v["method"].(string)
+			_, isTLS := v["tls"]
+			var h = "http"
+			if isTLS {
+				h = "https"
+			}
+			hr.Request.URL = &url.URL{
+				Scheme:  h,
+				Host:    v["host"].(string),
+				RawPath: v["uri"].(string),
+				Path:    v["uri"].(string),
+			}
+			if t, ok := m["bytes_read"]; ok {
+				hr.RequestSize = int64(t.(float64))
+			}
+			if t, ok := m["status"]; ok {
+				hr.Status = int(t.(float64))
+			}
+			if t, ok := m["size"]; ok {
+				hr.ResponseSize = int64(t.(float64))
+			}
+			if t, ok := m["duration"]; ok {
+				hr.Latency = time.Duration(t.(float64) * float64(time.Second))
+			}
+			hr.Request.Proto = v["proto"].(string)
+			hr.RemoteIP = v["remote_ip"].(string) + ":" + v["remote_port"].(string)
+
+			if t, ok := v["headers"]; ok {
+				headers := t.(map[string]any)
+				for h, v := range headers {
+					for _, s := range v.([]any) {
+						hr.Request.Header.Add(h, s.(string))
+					}
+				}
+			}
+			entry.HTTPRequest = &hr
+			//Caddy request contains more data, don't
+			//delete.
+			//delete(m, "request")
+		}
+		if val, exists := m["traceID"]; exists {
+			entry.Trace = "projects/" + l.projectID + "/traces/" + val.(string)
+			delete(m, "traceID")
+		}
+	}
+}
+
```

Comment:
Same idea as my comment in #9, casting should be done receiving and using the second optional param from the cast operation.